Add SSHManager support for invoking Windows workers via cmd.exe

mgkuhn opened the pull request:
Distributed.addprocs() now supports two new keyword arguments `ssh`
and `shell`.

Specifying `shell=:wincmd` now makes it possible to start workers on a
Windows machine with an sshd server that invokes cmd.exe as the shell
(as e.g. Microsoft's OpenSSH port does by default). Previously
SSHManager only supported ssh connections to a POSIX shell.

Specifying `ssh="/usr/bin/ssh"` makes it possible to specify the ssh
client that SSHManager will use (useful for debugging and where a
custom-version of ssh is required).

mgkuhn committed:
Add SSHManager support for invoking Windows workers via cmd.exe

Distributed.addprocs() now supports two new keyword arguments `ssh`
and `shell`.

Specifying `shell=:wincmd` now makes it possible to start workers on a
Windows machine with an sshd server that invokes cmd.exe as the shell
(as e.g. Microsoft's OpenSSH port does by default). Previosly
SSHManager only supported ssh connections to a POSIX shell.

Specifying `ssh="/usr/bin/ssh"` makes it possible to specify the ssh
client that SSHManager will use (useful for debugging and where a
custom-version of ssh is required).

mgkuhn committed:
escape Windows cmd.exe metacharacters with ^

mgkuhn commented on stdlib/Distributed/src/managers.jl:
```diff
@@ -184,29 +195,61 @@ function launch_on_machine(manager::SSHManager, machine, cnt, params, launched,
+        # shell login (-l) with string command (-c) to launch julia process
+        remotecmd = shell_escape_posixly(`sh -l -c $cmds`)
+
+    elseif params[:shell] == :wincmd
```
Fixed

mgkuhn commented on stdlib/Distributed/src/managers.jl:
```diff
@@ -225,6 +268,68 @@ function launch_on_machine(manager::SSHManager, machine, cnt, params, launched,
 end
 
 
+function print_shell_escaped_windows(io, args::AbstractString...)::Nothing
```
I have now also implemented an actual `cmd.exe` escaping algorithm on top of C-runtime ARGS escaping. The method I finally found (after several attempts) may seem a bit odd, but seems to work well: inserts a `^` escape character before any potential `cmd.exe` metacharacter whenever there was so far an even number of double quotes. This algorithms is needed because `cmd.exe` only half-treats double quotes as metacharacters: it does not remove them, but it uses them to toggle a quoting-mode flag, which in turn activates or deactivates other metacharacters, including `^`.

mgkuhn replied:
I have tested this patch for far only in the case where a Linux leader machine invokes a worker on a Windows machine (my use case is remotely controlling from Linux some laboratory kit for which only a Windows driver DLL is available). I have not yet tested remote method calls from Windows to Windows or form Windows to Linux, but I wouldn't expect there any changes needed. In particular, I would not expect any need for setting `windows_verbatim=true`. Without this flag, libuv already applies locally the same C-runtime escaping algorithm to the ssh commandline that I also had implemented for the remote command line, and that is needed to pass remotecmd on to the ssh process on Windows.

mgkuhn replied:
I've followed the function signature and naming of shell_escape_posixly() closely in case someone wants to put this into shell.jl, but I don't have particular views on what it should be called or where it should go.

vtjnash replied:
I think that's the clearest description of the cmd.exe algorithm I've ever heard.

I think it might be better to split this into a `shell_escape_win32ly` and `shell_escape_dosly` pair though, since some uses might only want one or the other (well, I don't care much about the exact names, just something memorable enough). I also think these are appropriate utility functions for inclusion in shell.jl. Can you move this there and add some tests? (you can probably just add them alongside the `shell_escape_posixly` tests)

musm replied:
According to my (potentially flawed) understanding, all the following tests should pass with this function but not all do
```julia
   @test Base.shell_escape_windows("a b c", "d", "e") == "\"a b c\" d e"
   @test Base.shell_escape_windows("ab\"c", "\\", "d") == "ab\\\"c\" \"\\\\\" d"
   @test Base.shell_escape_windows("a\\\\b", "d\"e",  "f\"g", "h") == "a\\\\\\b d\"e f\"g h"
   @test Base.shell_escape_windows("a\"b", "c", "d") == "a\\\"b c d"
   @test Base.shell_escape_windows("a\\\\b c", "d", "e") == "a\\\\\\\\\"b c\" d e"
```
(from https://docs.microsoft.com/en-us/cpp/c-language/parsing-c-command-line-arguments?view=vs-2019)

mgkuhn commented on stdlib/Distributed/src/managers.jl:
```diff
@@ -184,29 +195,61 @@ function launch_on_machine(manager::SSHManager, machine, cnt, params, launched,
+
+        any(c -> c == '"', exename) && throw(ArgumentError("invalid exename"))
+
+        remotecmd = shell_escape_windows(exename, exeflags...)
```
Are you sure? Seems to iterate fine:
```
julia> [a for a in `a b`]
2-element Array{String,1}:
 "a"
 "b"
```

vtjnash replied:
Ah, I forgot about 2211e0314eb (and did something stupid trying to test it). Yep, seems to be good.

vtjnash commented on stdlib/Distributed/src/managers.jl:
```diff
@@ -184,29 +195,61 @@ function launch_on_machine(manager::SSHManager, machine, cnt, params, launched,
+            remotecmd = "pushd \"$(dir)\" && $remotecmd"
+        end
+        if tval !== nothing && tval != ""
+            all(isdigit, tval) || throw(ArgumentError("invalid JULIA_WORKER_TIMEOUT"))
```
can we move this check outside the `if` block (and drop the escaping for the posix case)?

although this is too strict (the value is Float64-parsable)

vtjnash commented on stdlib/Distributed/src/managers.jl:
```diff
@@ -184,29 +195,61 @@ function launch_on_machine(manager::SSHManager, machine, cnt, params, launched,
+        cmdline_cookie = true
+        exeflags = `$exeflags --worker=$(cluster_cookie())`
+
+        any(c -> c == '"', exename) && throw(ArgumentError("invalid exename"))
```
should be safe (because of shell-escape below), although true it's not a valid Win32 name.

mgkuhn commented on stdlib/Distributed/src/managers.jl:
```diff
@@ -184,29 +195,61 @@ function launch_on_machine(manager::SSHManager, machine, cnt, params, launched,
+        remotecmd = shell_escape_windows(exename, exeflags...)
+        if dir !== nothing && dir != ""
+            any(c -> c == '"', dir) && throw(ArgumentError("invalid dir"))
+            remotecmd = "pushd \"$(dir)\" && $remotecmd"
```
I can't use `cd` because `cmd.exe` does not support setting the current working directory to a UNC path `\\filer\share\...`. I tried at first to get around this by changing the working directory via `julia -e "cd(\"\\\\filer\\share\...\")` only to realise that julia does not execute both options `-e` and `--worker` at the same time. I finally discovered that `pushd` is a much better command than `cd`, because when you give `pushd` a UNC, it automatically maps the share's root to a network drive as needed, and then `cmd.exe` happily takes that as its current working directory. (That discovery also reduced somewhat the need for very comprehensive string escaping that using option `-e` would have required, but having `cmd.exe` escaping implemented properly might still become useful one day if we put more stuff into `exeflags`.)

mgkuhn replied:
Since Windows does not permit `"` in filenames, there is no need to escape any of them in the argument of `pushd`. Therefore, always quoting the pathname passed to `pushd` seemed sufficient here.

mgkuhn replied:
"dosly" would be inaccurate because `cmd.exe` came new with Windows NT and is different from MS-DOS's `COMMAND.EXE`. (I don't recall if MS-DOS also had `^` escaping in `COMMAND.EXE`, it has been ~18 years since the last MS-DOS release ...)

vtjnash replied:
Ah, true. Wikipedia mentions that capability as an improvement of cmd.exe over COMMAND.COM. Even so, there may or might not be another special rule (for '%'): https://en.wikibooks.org/wiki/Windows_Batch_Scripting#Quoting_and_escaping

mgkuhn commented:
Self-reminder to test implications of recent OpenSSH quoting changes: https://github.com/PowerShell/Win32-OpenSSH/issues/1082

musm commented:
@mgkuhn would be great to have this in some form

musm commented on stdlib/Distributed/src/managers.jl:
```diff
@@ -225,6 +268,82 @@ function launch_on_machine(manager::SSHManager, machine, cnt, params, launched,
+                    write(io, '\\')
+                end
+                backslashes = 0
+                isword(c) || quoted || write(io, '^')   # for cmd.exe
```
@mgkuhn  Could you explain this logic here with `isword`?  
For the `isword` function it looks like  every cmd.exe special meta-character is excluded except '"' and '%'. 

What's confusing is that the `^` escaping depends upon whether the argument is `quoted`, why? 

In the below example `$` is quoted and is not `isword`
 `shell_escape_win(`string\"\$`) |> print` gives `"string\"^$"`

but in this example this line is not triggered
 `shell_escape_win(`string\$`) |> print` gives `"string$"`

What's the rationale here? This is the only major thing I don't understand in the code. Everything else correctly implements the inverse of `CommandLineToArgvW `

mgkuhn replied:
Welcome to [the wondrous world of the CMD.EXE parser](https://stackoverflow.com/questions/4094699/how-does-the-windows-command-interpreter-cmd-exe-parse-scripts). Whether `^` is an escape character or not depends on whether there have been an even or odd number of `"` on the line so far (see the Phase 2 description and the description of the quote flag in the previous link). We have not much control over the placement of `"`, as CMD.EXE always passes these on to the application, which then (this is typically done by libc according to the [conventions of the Microsoft C runtime](https://docs.microsoft.com/en-us/cpp/c-language/parsing-c-command-line-arguments)) uses them to split the command-line string into an argv array. Therefore we have to alternate between the two available quoting mechanisms. This all gets complicated by the fact that the `"` are seen and interpreted by both CMD.EXE and the application.

musm replied:
@mgkuhn  Thanks for the reply. (BTW I opened a PR to get portions of this PR merged, only because I thought you  weren't interested in pushing this PR through due to inactivity. I'd be glad if you have the time otherwise, I can try to understand the issue and push it through.)

I think I kind of understand the gist of your comment. However wouldn't it be much easier to implement the following logic described at the end of the post here.  https://blogs.msdn.microsoft.com/twistylittlepassagesallalike/2011/04/23/everyone-quotes-command-line-arguments-the-wrong-way/

> 1 Always escape all arguments so that they will be decoded properly by CommandLineToArgvW, perhaps using my ArgvQuote function above.
> 
> 2 After step 1, then if and only if the command line produced will be interpreted by cmd, prefix each shell metacharacter (or each character) with a ^ character.
> 

Currently your code tries to intelligently infer if we should quote the argument depending if all the characters in the word are `isword`. Since we know each argument  is the output of `shell_split`, it doesn't matter  and we can just `\"arg\"`. This is much simpler. 

As consequence of step 1, we will always have an even number of `"`  and thus `^` will always be an escape character.

musm replied:
In any case we can disregard my comment above, since it does indeed seem simpler to use your idea to toggle the state depending on `"` .  It seems that all the other processing should be properly taken care of by CMD, other than handling `^`.

mgkuhn replied:
I'm still planning to push some variant of this PR through by the end of this year, but I will need more time for thorough testing as well as also covering PowerShell. I'm nore sure this PR is actually closely related to the one of Shell mode in REPL. The escaping routines in this PR are designed to pass strings transparently (past CMD.EXE or PowerShell) to an application (such as Julia) that parses the command line using the Microsoft C conventions. In REPL Shell mode, this is not the only or even main requirement. You may want to send commands that are interpreted by CMD.EXE, which parses its lines using completely different conventions from CommandLineToArgvW. I don't see how any function inverting CommandLineToArgvW would be of use in REPL Shell mode for Windows, as CMD.EXE does no use anything remotely resembling CommandLineToArgvW. I haven't looked yet closely into how REPL Shell mode works, but for Windows it would have to pass the whole command line transparently as a single string to the shell.

mgkuhn committed:
adding escape_raw_string()

Implementation of the raw-string escaping convention. Can also be
used to escape command-line arguments for Windows C/C++/Julia
applications, which use the same \" escaping convention as
Julia non-standard string literals.

mgkuhn committed:
refactored shell_escape_winsomely() into escape_microsoft_c_args()

Using escape_raw_string() make this function much easier to
understand. Name changed, as the function is not related to any shell.

mgkuhn commented:
> I think it might be better to split this into a `shell_escape_win32ly` and `shell_escape_dosly` pair though, since some uses might only want one or the other (well, I don't care much about the exact names, just something memorable enough). I also think these are appropriate utility functions for inclusion in `shell.jl`. Can you move this there and add some tests? (you can probably just add them alongside the `shell_escape_posixly` tests)

@vtjnash Sorry for the long delay, I have now done all of that in #34111

mgkuhn committed:
fix dependency problems, load regex.jl before shell.jl

mgkuhn committed:
shell_escape_wincmd(): pass through " only in pairs

* ensure that only an even number of double quotes is passed through
  unescaped, such that the quote flag in CMD.EXE remains unset afterwards

* clarify in documentation why no attempt to escape `%`

* operates now on bytes rather than Unicode characters (to avoid
  UTF-8 decoding overhead)

* @ is now escaped at start of string

* more optimizations and test cases

mgkuhn committed:
fixed doctest

mgkuhn committed:
remove shell_escape_winsomely()

mgkuhn committed:
replace shell_escape_windows with shell_escape_wincmd and escape_microsoft_c_args from PR #34111

mgkuhn committed:
addprocs: new keyword arguments `env` and `cmdline_cookie`

* generalized mechanics for passing environment variables to workers

* `cmdline_cookie` is work around for an ssh problem with Windows workers that run older (pre-ConPTY) Windows or Julia versions

mgkuhn committed:
shell_escape_wincmd: use proper AbstractString API

No longer assume that AbstractString uses some ASCII compatible encoding
(such as UTF-8), but call the UTF-8 decoder for each character
(even though this causes 30-60% runtime overhead compared to the
previous version).

mgkuhn commented:
Now rebased on top of the latest commit on #34111, the separated out PR that adds shell_escape_wincmd().
What else is needed before this can be merged?

ViralBShah commented:
Bump. Apart from rebasing, what else do we need to get this in?

mgkuhn commented:
This (basically relatively simple) PR got unfortunately dragged out a bit, probably unnecessarily. It initially contained some shell-escaping functions to prepare Windows command lines, which @vtjnash suggested are more widely useful beyond Distributed and he therefore suggested these should go into base/shell.jl instead. There are really two functions: one escaping metacharacters of CMD.EXE, the other formatting a command line such that arbitrary characters can be passed on to a Microsoft C program's argv array. (Unlike Unix, Windows passes on a command line to an application as a single string, not as an array of strings, therefore a C program has its own parser to split that into argv[], and we need an inverse of that function.) Then @musm added one of these functions, the one that prepares an argv[] string, but using the rather odd name `shell_escape_winsomely()`, odd because it does not actually escape any “shell” functionality but Microsoft libc functionality. I then prepared #34111, which refactors and renames @musm's shell_escape_winsomely() into escape_microsoft_c_args(), and also adds the second function as shell_escape_wincmd().

Therefore my suggestion is to first look at #34111, and once that is merged, this PR here should be pretty straight forward.

I'll prepare another rebase of #34111for you to look at.

mgkuhn commented:
Replacing this PR with PR #30614.
